spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java

```diff
@@ -303,52 +309,71 @@ protected boolean shouldCopyRequestHeaders() {
 
 		private final MessageHandler delegate;
 
+		private final boolean supportsSerializationNatively;
+
 		private SendingHandler(MessageHandler delegate, boolean embedHeaders,
-				String[] headersToEmbed) {
+				String[] headersToEmbed, boolean supportsSerializationNatively) {
 			this.delegate = delegate;
 			this.setBeanFactory(AbstractMessageChannelBinder.this.getBeanFactory());
 			this.embedHeaders = embedHeaders;
 			this.embeddedHeaders = headersToEmbed;
+			this.supportsSerializationNatively = supportsSerializationNatively;
 		}
 
 		@Override
 		protected void handleMessageInternal(Message<?> message) throws Exception {
-			MessageValues transformed = serializePayloadIfNecessary(message);
-			byte[] payload;
+			MessageValues messageToSend = (this.supportsSerializationNatively) ? new MessageValues(message) :
+					serializePayloadIfNecessary(message);
+			this.delegate.handleMessage(embedHeadersIfApplicable(messageToSend));
```

Comment:
Again, I think this can be simplified and the old logic can be restored as there should be no change to it - we just need to add a separate case where native encoding is being used - then we just pass the object to the binder.